### PR TITLE
feat(oauth): force validate provided redirect uri

### LINKED ISSUE #5827: Clean up redirect_uri management for scoped clients
- need to double check the trailing slash logic
- more

### PR DESCRIPTION
**Pull request by vladikoff**
BREAKING CHANGE: OAuth redirect uris must be updated in the database

@rfk r?

Fixes #5827

### CODE REVIEW

File: app/scripts/models/reliers/oauth.js [hunk 1]
```diff
@@ -170,13 +170,22 @@ define(function (require, exports, module) {
...
+      // get the app provided redirect uri
+      const queryRedirectUri = this.get('redirectUri');
 
       return this._oAuthClient.getClientInfo(clientId)
         .then((serviceInfo) => {
-          var result = Transform.transformUsingSchema(
+          const result = Transform.transformUsingSchema(
             serviceInfo, CLIENT_INFO_SCHEMA, OAuthErrors);
+
+          // verification flow doesn't have a redirect uri, so there is nothing to validate
```
Comment (vladikoff) [inline app/scripts/models/reliers/oauth.js]: @rfk thoughts on this plot twist?

Comment (rfk) [reply app/scripts/models/reliers/oauth.js]: This wasn't clear from the discussion in #5827 but IIUC, clients are not *required* to provide the redirect_uri in the query parameters, but *if* they provide it then we should validate it and error out on mismatch.  So the logic would be something like:

```
if (queryRedirectUri && queryRedirectUri !== result.redirectUri) {
  throw OAuthErrors.toError('INCORRECT_REDIRECT');
}
```

Which should work OK with the verification flow where it's not provided?

Or, am I misunderstanding something that means clients *must* provide the redirect_uri in the query string for other reasons?

**Comment by rfk** [comment]
> BREAKING CHANGE: OAuth redirect uris must be updated in the database

What's the sequence for rolling this out, do we simply need to ensure the URLs are updated in the DB before this change goes to prod?

File: tests/functional/oauth_query_param_validation.js [hunk 2]
```diff
@@ -7,16 +7,17 @@
...
+const config = intern._config;
 
-var SIGNUP_ROOT = config.fxaContentRoot + 'oauth/signup';
-
-var TRUSTED_CLIENT_ID;
-var TRUSTED_SCOPE = 'profile';
-var UNTRUSTED_CLIENT_ID;
-var UNTRUSTED_SCOPE = 'profile:uid profile:email';
-var UNTRUSTED_NO_VALID_SCOPES = 'profile';
+const SIGNUP_ROOT = config.fxaContentRoot + 'oauth/signup';
```
Comment (vladikoff) [inline tests/functional/oauth_query_param_validation.js]: \` \` in here

**Comment by vladikoff** [comment]
> do we simply need to ensure the URLs are updated in the DB before this change goes to prod?

Run db query to add `/` to redirect uris then flip content server to latest version after that.

File: app/scripts/models/reliers/oauth.js [hunk 3]
```diff
@@ -171,12 +171,26 @@ define(function (require, exports, module) {
...
+          const result = Transform.transformUsingSchema(
             serviceInfo, CLIENT_INFO_SCHEMA, OAuthErrors);
+
+          /**
+           * If redirect_uri was specified in the query we must validate it
+           * Ref: https://tools.ietf.org/html/rfc6749#section-3.1.2
+           *
+           * Verification (email) flows do not have a redirect uri, nothing to validate
+           */
+          if (queryRedirectUri && result.redirectUri !== queryRedirectUri) {
```
Comment (vladikoff) [inline app/scripts/models/reliers/oauth.js]: @rfk updated!

**Comment by vladikoff** [comment]
> are all the test changes to include the redirect_uri still needed?

I wanted to include them because that's the "real" behaviour. I think most of the oauth reliers send the redirect_uri today.

**Comment by vladikoff** [comment]
> > are all the test changes to include the redirect_uri still needed?

> I wanted to include them because that's the "real" behaviour. I think most of the oauth reliers send the redirect_uri today.

If that feels like the wrong thing to do I can undo the changes.

**Comment by rfk** [comment]
No, it's fine as-is, thanks for clarifying.

**Comment by rfk** [comment]
> it's fine as-is

Erm, in case that was ambigious - the PR is fine as-is with all the test changes included, I just wanted to make sure I understood what they were for :-)

**Pull request merged**